### ISSUE TITLE
User-defined Folders/Categories for the 'Mods' tab

### ISSUE THREAD

**Issue by JLUsr** [PrismLauncher/PrismLauncher#154]
### Role

I use Prism Launcher to troubleshoot and play modded Minecraft

### Suggestion

I think it would be great if the `Mods` tab within an instance had the ability to sort/categorize files into user-defined categories

### Benefit

This would make troubleshooting incompatibilities and crashes a lot easier.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

I have seen something similar listed on MultiMC's GitHub page and mentioned on PolyMC's so I thought it would be nice to make something visible here. The idea is to create a sub-folder/category within the `Mods` tab as a way to easily manage mods. It would be great for troubleshooting potential conflicts since it would make separating required/library mods less of a hassle for those who want to go with the 'Take out half your mods and see if it works' method without hunting down individual mods when you accidentally remove a dependency or constantly trying to make sure you haven't accidentally removed a dependency too early.

Made some quick examples, they aren't the best but it gets the point across.
![Enabled](https://user-images.githubusercontent.com/47650360/196838604-cb59e977-04f2-4ef9-9223-826743922a1c.png)
![EnabledDropDown](https://user-images.githubusercontent.com/47650360/196838620-8dbb60ec-3162-41d5-9581-fe975901f302.png)

Additionally, an example where one of the categorized mods is disabled.
![PartiallyEnabled](https://user-images.githubusercontent.com/47650360/196838746-cb688359-cd84-42e3-8a9b-39d80ffbab06.png)
![PartiallyEnabledDropDown](https://user-images.githubusercontent.com/47650360/196838752-24656640-51e8-4cab-8a32-e0f6883e2958.png)

P.S. I was too lazy to number them like `ExampleMod1` and `ExampleMod2` so sorry about that.

**Comment by microlith57** [comment]
would recommend having a tristate checkbox to distinguish between every mod in a category being disabled and just one

**Comment by JLUsr** [comment]
> would recommend having a tristate checkbox to distinguish between every mod in a category being disabled and just one

The example in which "...one of the categorized mods is disabled." was intended to represent a new third state that would only apply to categories where there is a mix (in any amount) of mods being enabled and disabled within a specific category. By no means was I intending for that to be viewed as the default when every mod is disabled. The goal would be to have the default `Enabled` and the default `Disabled` states represented only when **every** mod within a category is, respectively, enabled or disabled. I was just too lazy to make another example in which every mod in a category was represented as being `Disabled`.

Additionally, in the event you were concerned that you wouldn't be able to differentiate easily between a category being disabled (and subsequently all of the mods contained in that category being disabled) and an uncategorized mod being disabled, I would hope that the appearance of an icon used to expand/minimize a category would be enough of a representation that what is disabled is either a category or solitary mod. That being said, this brings rise to the idea that having a margin between categories/folders and individual mods might be nice, sectioning them off from each other slightly so that sorting by Dates or Enabled/Disabled state could still be done without mixing categories of mods and solitary mods.

**Comment by microlith57** [comment]
ah, i guess i'm used to the `[✓]` 'all' / `[-]` 'some' / `[ ]` 'none' icon scheme, i didn't parse `[■]` as meaning 'some'

**Comment by Aces-and-Jacks** [comment]
Is this still being considered?

**Comment by JLUsr** [comment]
> Is this still being considered?

It's safe to say that it hasn't been denied considering they didn't close the issue themselves. It would be best to assume they are focusing on fixes and updates regarding stability. I assume it will be properly addressed when they have time to focus on QOL updates.

**Comment by Aces-and-Jacks** [comment]
So maybe 7.0 hopefully. Whenever they decide to do this

**Comment by TheAyes** [comment]
I can add on to this:
Quilts Version 0.18 Modloader actually supports sub-folders in mods. When done however it only shows these folders in prism which is sub-optimal for deactivating specific mods.
So I can 100% agree with wanting such a feature!

**Comment by Aquaeyes** [comment]
Not having to move mods out of their folders every time I update would be wonderful ^-^'

**Comment by mart3323** [comment]
To add another use case that i keep running into.

Whenever i play a modpack, i usually want to add some extra mods myself (item scroller, inventory+, tweakeroo, etc.). 
But when i do so, then if i run into a mod conflict i have to scroll through the entire mod list and try to remember which mods are part of the pack and should be left alone, and which are the ones I added.
The only way to distinguish is if i create a second instance of the modpack and open that on the side to compare

With mod groups, i could just put all the modpack mods into one group before i add any of my own, then make sure to put all the ones i add later in another
Now if there's an issue i only have to check the group with my added mods (and can even easily do a binary search in it)

**Comment by BoardAnims** [comment]
Since Prism has a system for automatically downloading a mod's dependency(ies), I feel like it wouldn't be too hard to isolate any/all detected dependencies when batch disabling mods.

**Comment by cintreck** [comment]
please, i will use this launcher over atlauncher if you add this feature

**Comment by Jakub-Wilk** [comment]
I'm currently playing a Forge modpack, but I sometimes want to use Axiom when building in a creative testing world. To use Axiom, I have to enable Sinytra Connector (as Axiom is a Fabric mod), enable Axiom, and disable Embeddium and its dependants (incompatibility). Searching for those mods in the 250+ mod list and toggling them one by one every time I want to switch from building to playing survival is very annoying, having the capability to just put those mods into groups and toggle the groups would be extremely useful to me.

**Comment by 3meraldK** [comment]
I totally support this idea. If you have plenty of mods, this feature would come handful. But as for my personal and additional request - make it real for resource packs, shader packs, worlds, servers and screenshots as well.

**Comment by danaYatsuta** [comment]
Seconding this request, would be very great to have

**Comment by Kationor** [comment]
With #2645 being marked as a duplicate, I do want to address a fundamental question: Tags or Folders?

Folders might be the simpler option, though they give users less customisability. There'd also be the choice between representing them as file system directories and launcher metadata. File system directories would be more portable (users can share their mod directory with people who don't use Prism more easily) but they would not be supported by all modloaders, especially on older versions. Launcher metadata would keep modloader compatibility, but reduce the portability between launchers.

Tags would give users more freedom, though full support would be more complex. Tags are not easily representable on the file system, so would need to be launcher metadata. This also means that they would be compatible with all modloaders. To get the most out of tags, the current filter system would need a major upgrade, but this would not be needed for initial support. Initial support should at least consist of the ability to add user-defined tags to specific mods and the ability to select a group of mods that share a tag.

I am personally in favour of basic tag support, with the possibility of expansion at a later date if tags become widely used.

**Comment by Trial97** [comment]
it will be tags 100%(folders may not be compatible with loaders)

**Comment by teacat923** [comment]
I am the author of  #2645  and I think tags are easier to manage as there is more information that can be attached to one mod.
Most importantly, it is superior in that it does not destroy the existing mod management system.

**Comment by teacat923** [comment]
I am the author of  #2645 and I think tags are easier to manage as there is more information that can be attached to one mod.
Most importantly, it is superior in that it does not destroy the existing mod management system.

**Comment by JLUsr** [comment]
> it will be tags 100%(folders may not be compatible with loaders)

This post was **not** intended to be a way to make folders on the drive for storing jar files. It's more like categories for virtual organization. The goal is to sort things in Prism not to sort them in the hard drives directory. Personally, I don't really want anything to move where my jar files are stored in my drive. It seems I didn't express this too well in my initial post but I hope this makes my intent a little more clear.

> I am the author of #2645 and I think tags are easier to manage as there is more information that can be attached to one mod. Most importantly, it is superior in that it does not destroy the existing mod management system.

I think #2645 is not really a duplicate of this issue, as organization via tags is a different thing. You can have both tags and Folders/Categories work together for more detailed organization and flexibility. The reason this post is about Folders/Categories specifically is because I felt it matched my individual use case more accurately than tags would. I still feel that way; however, I also like the idea of tags. Heck, I even want to use tags to organize Folders/Categories even further now that I really think about it. In the end, I hope both are implemented.

I also want to state that my suggested method would not "destroy the existing mod management system" since it wasn't in anyway intended to reference a change to the way files are stored or managed on the drive. One example of this might be the way ModOrganizer2 chose to do their separators (I would prefer a different form of executing this feature compared to the way it is done in MO2. Which is why I chose to use the term Folders/Categories rather than the term Separator and why I provided the examples in the initial post). Edit: _I considered removing this part since I don't actually know the technical details relevant to the existing mod management system and commented with my perspective being focused on the storage of the actual jar files. In the end I kept it because I still felt it was worth it to drive that point home; however, I do want to make it clear that, since I lack the knowledge on this specific subject, I could totally be wrong about this. It just seems reasonable that the choice in implementation is what determines how destructive a feature might end up being._

I would like to have this issue be treated as separate if at all possible. I am all for improving and modifying someone's idea within a certain scope; but, I also think that it's rather unfair to treat ideas that are intended to accomplish similar goals as being the same idea. We all want to have a way to manage and organize our mods within the launcher; but, we might not agree on the method used to actually accomplish that feat.

(To be clear, I am in no way trying to slight anyone here with what I had to say. I can also see how the above might come off as being a "Choosy Beggar" but I guess all I can do is hope that no one views it as such)

**Comment by Kationor** [comment]
> @JLUsr: I also think that it's rather unfair to treat ideas that are intended to accomplish similar goals as being the same idea.

In this case, I do think that they are effectively the same idea. The way I see it is that folders, when represented as launcher metadata, can just be considered mutually exclusive tags. If a user still wants folders, they can simply elect to only apply a single tag to each mod. This means that a general system that handles tags could be extended to folders with a bit of UI work (implement a "folder view" that treats each tag as a folder and the tagged mods as its children).

**Comment by Kationor** [comment]
So while I do think that tags are the way to go, I have other implementation questions:
- Should you be able to nest folders (or tag tags with tags)?
- How would filtering by tags look? Would the current text filter be extended (might require parsing) or would it be a separate UI?

**Comment by Aquaeyes** [comment]
(Perhaps a toggle for switching between virtual folders and actual folders could be good? That may be a bit too much bloat, but I do actually use folders and nested folders in my setups) - though a virtual system would work nicely for the same usecases. The interpretation I had earlier was centered around the fact that I already do have folders in my packs, but Prism is unable to see anything but the folder itself.

**Comment by WGINC** [comment]
> So while I do think that tags are the way to go, I have other implementation questions:
> 
> * Should you be able to nest folders (or tag tags with tags)?
> * How would filtering by tags look? Would the current text filter be extended (might require parsing) or would it be a separate UI?

First of all, bump. Second of all, I'm partial to allowing the nesting of folders/tags as long as it is only changing the GUI side of things. Lastly, I would advise that the mods should be tagged; but the tags be displayed to the user as folders. 

e.g. mod A has the tags (either given from the mod maker or the user): combat, worldgen, and utility. So, to the user it would be placed in the folders: combat, worldgen, and utility. If you enable the mod in the worldgen folder, it enables the mod in the combat and utility folders. And by extension, if you enable the mod folder worldgen, the mod enables itself in the other two folders.

**Comment by teacat923** [comment]
Is this project still working?

**Comment by Trial97** [comment]
> Is this project still working?

No maintainers are working on this as of yet.
But yes this is planed.
So if you want to contribute feel free to do so.

**Comment by Aces-and-Jacks** [comment]
I don't know if it's been said but what about a `json` file in the mods folder that would visually sort everything, but the files in the folder wouldn't be touched. Shouldn't be the same as how instances are set up. Since to me instances are bigger and their own group, but in the mods folder it'd be easy to find the mods and would be easier to sort if a json file managed it. Similar to how ferium sorts it

**Comment by Hikiru** [comment]
i want to use the cobblemon official modpack as a base for my modpack. there's already a ton of optimization and small QOL mods that it immediately becomes a hassle trying to keep track of what i add. i feel like mod organization should be a big priority. prism is primarily for easy modding, but creating your own modpack with it is very annoying without any organization tools

**Comment by InfiniteCoder01** [comment]
As a workaround, you can put mod JARs into separate directories and add "-Dfabric.addMods=mods" to java options (for Fabric, couldn't find anything for Forge, but might exist). Of course, this totally breaks PrismLauncehr, as it's unable to find mods in subfolders (in fact, it recognizes subfolders as mods themselves).

**Comment by radishcatish** [comment]
i Need this so badly

**Comment by greensad** [comment]
Easily the biggest killer feature that this loader is missing, really do hope we get some form of this someday since the lack of mod compartmentalization really slows down the process

**Comment by buhhhhg** [comment]
is this still being considered?

**Comment by Trial97** [comment]
Yes it is, but there are a few technical stuff that need to be fixed first:
- a place where to store that information about the mods
- how should it be displayed to the user

Right now no maintainer is working on this, so any contributor is welcomed.
(Any open GitHub issue is considered)

**Comment by abhicommands** [comment]
I have found that currently in the latest 10.0.0 pre release you can just add the "-Dfabric.addMods=mods" to jvm arguments and run sub folders, however those subfolders inside the mods aren't displayed to users or upgradable, I am currently working on a feature to make this so that those subfolders are displayed, but there will still not be a way to actually run this automatically without putting the jvm argument "-Dfabric.addMods=mods" manually even after adding the support for subfolders users must know that they have to add "-Dfabric.addMods=mods" manually for their mods to work from a sub folder. 

The other feature that is really helpful is if modpacks specifically fabric modpacks, are installed in a folder that way when you update ur custom mods that aren't from the modpack it works without interfering with the modpack, this is a special feature thats also something I am working on. Will be submitting a PR will see if it gets into this new release!


personally I think that now in the next release whenever someone adds a new instance and its a fabric modpack from cursgeforge or modrinth it just automatically puts in a folder for easier stuff management but I am not too sure so we can just have a prompt before starting the installation to say do u wanna install modpack in a folder instead and if so it adds the -Dfabric.addMods=mods automatically to jvm arguments.

**Comment by ErrorUnseen** [comment]
@abhicommands your pull #4571 states it fixes this issue but it is only for fabric right? it adds support for the fabric mod loaders ability to load mods from sub folders in the mods folder. that is different from this issue right? cool feature to have but only for those on fabric, does this mean that forge users will be getting these prompts for modpacks that would break things right?

**Comment by abhicommands** [comment]
> [@abhicommands](https://github.com/abhicommands) your pull [#4571](https://github.com/PrismLauncher/PrismLauncher/pull/4571) states it fixes this issue but it is only for fabric right? it adds support for the fabric mod loaders ability to load mods from sub folders in the mods folder. that is different from this issue right? cool feature to have but only for those on fabric, does this mean that forge users will be getting these prompts for modpacks that would break things right?

@ErrorUnseen Nope, thats not the case, I have made sure that this prompt is only going to be there for modpacks that are fabric and the setting is only available for fabric, but you are getting things, wrong, what I am adding isn't a deliberate feature for this, I am only adding a way to see stuff thats also inside folders. Right now whats happening is that if there is a folder inside my mods folder, it doesn't show it at all and it doesn't work it breaks things, I added a sub tree folder based mod organizing support on the mods tab for instances, that way the instances can show mods and stuff inside sub folders as well, but that doesn't mean it fixes it it. I hope that helps! This technically only works for fabric because fabric is the only loader that allows for this support but we can't do anything about that, if other loaders also decide that they wanna support it then yea this would for for all!


A workaround, can be that there can be a mods-prism(organizer) where these mods can be organized and then copied as a whole this can be done for forge mods, Or another actual fix that can work for all modpacks, can be also implemented but the Prism devs have absolutely no interest in this stuff and aren't supporting me on this, is where the metadata and a psuedo organizer inside the prism launcher itself, where mods are organized inside the launcher itself instead of different folders that would make things a lot easier but thats not yet installed all u have to do for that is just update how you work the metadata and it will fix it but I am currently in the process of designing that we'll see how far it will go.

**Comment by Azyrod** [comment]
I really want this feature. I have currently more than 100 mods on my main instance and it is such a hassle to manage them properly. I also have some performance issues which I don't have with a more minimalistic instance, and it would be of immense help to be able to make a "Good" / "Bad" group for bisecting the mod(s) causing the issue.

Other than that, I am completely unable to locate what mods I have. I need to read all of the names, try to remember what they do when I make decisions about adding/removing mods. It would be amazing to be able to group them into categories like "Performance", "QOL", "Rendering", "Creative Tools", "Utilities" ect.

**Comment by abhicommands** [comment]
> I really want this feature. I have currently more than 100 mods on my main instance and it is such a hassle to manage them properly. I also have some performance issues which I don't have with a more minimalistic instance, and it would be of immense help to be able to make a "Good" / "Bad" group for bisecting the mod(s) causing the issue.
> 
> Other than that, I am completely unable to locate what mods I have. I need to read all of the names, try to remember what they do when I make decisions about adding/removing mods. It would be amazing to be able to group them into categories like "Performance", "QOL", "Rendering", "Creative Tools", "Utilities" ect.

I am currently trying to code it, if anyone here really wants these features then please you should check out pr #5033  and test to see if it fits ur requirements and send any bugs/issues u found while testing it which would aid development for me and the maintainers to see as well!

**Comment by notVivekG** [comment]
Hey @Azyrod @Jakub-Wilk  @mart3323  @WGINC @Kationor , 
Related to this issue, though using a different mechanism than the tags/folders direction being built in #5033. 
I've put together a Mod Profile/Tab system (named, switchable sets of enabled mods) for the Mods Page , more context over on #1770

**My PR is** #5719

One idea I'm weighing as a follow-up (not part of the current PR): letting you select **multiple profile tabs** at once, where the mods that actually run in-game are the union of whichever tabs are selected. So instead of just switching between **"Optimization"** and **"Building"** , you could combine something like "Base Pack" + "Performance Tweaks" + "My Testing Mods" as one merged loadout.

Wanted to check whether that's close to what people here have been asking for with groups, or whether it's solving a different problem. If there's interest I'll scope it as a proper follow-up once the base profile PR lands. Feedback welcome either way on the profile system itself or the merge idea.

**Comment by Azyrod** [comment]
@notVivekG I really like the idea and it nicely solves a couple of issues, including the one described here where we want to be able to debug a mof issue via bisection. It also solves the issue of having to create new instances all the time, when we just want to switch between 2 sets of mods

Being able to select multiple profiles might be nice and would allow to use theses profiles as categories / groups, but idk how easily that would be to implement (what would the UI even look like?)

I'll try to take a look during the week

**Comment by XymaScope** [comment]
Making a note of the active development at PR #5719 for people just searching this thread and not realising there's discussion in PRs directly too (me, i forgot that, i'm the silly one)

Please nuke this if thats not a good use of a comment, don't want to bump spam ^-^;
